openmmapi/src/NoseHooverIntegrator.cpp

```diff
@@ -0,0 +1,289 @@
+                                                   [&particle2](const std::pair<int, int>& pair){ return pair.first == particle2 || pair.second == particle2;})
+                                                      != thermostatedPairs.end());
+        if ((system.getParticleMass(particle1) > 0) && (system.getParticleMass(particle2) > 0)){
+            if ((particle1_in_thermostatedParticles && !particle2_in_thermostatedParticles) ||
+                 (!particle1_in_thermostatedParticles && particle2_in_thermostatedParticles)){
+                throw OpenMMException("Cannot add only one of particles " + std::to_string(particle1) + " and " + std::to_string(particle2)
+                                        + " to NoseHooverChain, because they are connected by a constraint.");
+            }
+            if (particle1_in_thermostatedParticles && particle2_in_thermostatedParticles){
+                nDOF -= 1;
+            }
+        }
+    }
+    int numForces = system.getNumForces();
+    if (thermostatedPairs.size() == 0){ // remove 3 degrees of freedom from thermostats that act on absolute motions
+        for (int forceNum = 0; forceNum < numForces; ++forceNum) {
+            if (dynamic_cast<CMMotionRemover*>(&system.getForce(forceNum))) nDOF -= 3;
+        }
+    }
+
+    // make sure that thermostats do not overlap
+    for (auto &other_nhc: noseHooverChains) {
+        for (auto &particle: thermostatedParticles){
+            bool isParticleInOtherChain = (std::find(other_nhc.getThermostatedAtoms().begin(),
+                                                     other_nhc.getThermostatedAtoms().end(),
+                                                     particle) != other_nhc.getThermostatedAtoms().end()) ||
+                                          (std::find_if(other_nhc.getThermostatedPairs().begin(),
+                                                        other_nhc.getThermostatedPairs().end(),
+                                           [&particle](const std::pair<int, int>& pair){ return pair.first == particle || pair.second == particle;})
+                                              != other_nhc.getThermostatedPairs().end());
+            if (isParticleInOtherChain){
+                throw OpenMMException("Found particle " + std::to_string(particle) + "in a different NoseHooverChain, "
+                                      "but particles can only be thermostated by one thermostat.");
+            }
+        }
+    }
+
+    // create and add new chain
+    int chainID = noseHooverChains.size();
+    auto nhc = NoseHooverChain(temperature, relativeTemperature, collisionFrequency, relativeCollisionFrequency,
+                               nDOF, chainLength, numMTS, numYoshidaSuzuki, chainID, thermostatedParticles, thermostatedPairs);
+    noseHooverChains.push_back(nhc);
+    return chainID;
+}
+
+double NoseHooverIntegrator::getTemperature(int chainID) const {
+    if (chainID >= noseHooverChains.size()) {
```

Comment:
All these range checks can be simplified to just

```C++
ASSERT_VALID_INDEX(chainId, noseHooverChains);
```

It also checks for negative indices.